man/genericups.8
.TH GENERICUPS 8 "Thu Aug 11 2005" "" "Network UPS Tools (NUT)"
.SH NAME
genericups \- Driver for contact\(hyclosure UPS equipment

.SH NOTE
This man page only documents the specific features of the genericups 
driver.  For information about the core driver, see \fBnutupsdrv\fR(8).

.SH SUPPORTED HARDWARE
This driver supports hardware from many different manufacturers as it only 
uses the very simplest of signaling schemes.  Contact closure refers to a 
kind of interface where basic high/low signals are provided to indicate 
status.  This kind of UPS can only report line power and battery status.

This means that you will only get the essentials in ups.status: OL, OB,
and LB.  Anything else requires a smarter UPS.

.SH CABLING
Cabling is different for every kind of UPS.  See the table below for 
information on what is known to work with a given UPS type.

.SH EXTRA ARGUMENTS
This driver supports the following settings in the \fBups.conf\fR(5):

.IP "upstype=\fItype\fR"

Required.  Configures the driver for a specific kind of UPS.  See the UPS 
Types section below for more information on which entries are available.

.IP "mfr=\fIstring\fR"

Optional.  The very nature of a generic UPS driver sometimes means that 
the stock manufacturer data has no relation to the actual hardware that is 
attached.  With the mfr setting, you can change the value that is seen by 
clients that monitor this UPS.

.IP "model=\fIstring\fR"

Optional.  This is like mfr above, but it overrides the model string 
instead.

.IP "serial=\fIstring\fR"

Optional.  This is like mfr above and intended to record the identification
string of the UPS. It is titled serial because usually this string is 
referred to as the serial number.

.IP "sdtime=\fIvalue\fR"

Optional.  The driver will sleep for this many seconds after setting the 
shutdown signal.  This is necessary for some hardware which requires a
sustained level to activate the shutdown sequence.

The default behavior of the driver is to exit immediately.  If this
doesn't reliably trigger a shutdown in your UPS hardware, use this
setting to give it more time to react.

Note: very large values for sdtime may create warnings from upsdrvctl if
it gets tired of waiting for the driver to return.

.SH CUSTOM CONFIGURATIONS

You may override the values for CP, OL, LB, and SD by defining them in
the \fBups.conf\fR after the upstype setting.

For example, to set the cable power to DTR and the low battery value to
DCD, it would look like this:

	CP = DTR

	LB = DCD

Recognized values for input lines are CTS, DCD, and RNG.  Recognized
values for output lines are DTR, RTS, and ST.  See below for more about
what these signals mean.

These values may be negated for active low signals.  That is, "LB=\-DCD"
recognizes a low battery condition when DCD is not held high.

.SH TYPE INFORMATION
The essence of a UPS definition in this driver is how it uses the serial 
lines that are available.  These are the abbreviations you will see below:

	OL \(hy On line (no power failure) (opposite of OB \(hy on battery)

 	LB \(hy Low battery

	SD \(hy Shutdown load

	CP \(hy Cable power (must be present for cable to have valid reading)

	CTS \(hy Clear to Send.  Received from the UPS.

	RTS \(hy Ready to Send.  Sent by the PC.

	DCD \(hy Data Carrier Detect.  Received from the UPS.

	RNG \(hy Ring indicate.  Received from the UPS.

	DTR \(hy Data Terminal Ready.  Sent by the PC.

	DSR \(hy Data Set Ready.  Received from the UPS.

	ST \(hy Send a BREAK on the transmit data line

A \- in front of a signal name (like \-RNG) means that the indicated 
condition is signaled with an active low signal.  For example, [LB=\-RNG] 
means the battery is low when the ring indicate line goes low, and that 
the battery is OK when that line is held high.

.SH UPS TYPES

0 = UPSonic LAN Saver 600 
    [CP=DTR+RTS] [OL=\-CTS] [LB=DCD] [SD=DTR]

1 = APC Back\(hyUPS/Back\(hyUPS Pro/Smart\(hyUPS with 940\-0095A/C cable 
    [CP=DTR] [OL=\-RNG] [LB=DCD] [SD=RTS]

2 = APC Back\(hyUPS/Back\(hyUPS Pro/Smart\(hyUPS with 940\-0020B cable 
    [CP=RTS] [OL=\-CTS] [LB=DCD] [SD=DTR+RTS]

    Type 2 has also been reported to work with the 940\-0020C cable.

3 = PowerTech Comp1000 with DTR cable power 
    [CP=DTR] [OL=CTS] [LB=DCD] [SD=DTR+RTS]

4 = Generic RUPS Model 
    [CP=RTS] [OL=CTS] [LB=\-DCD] [SD=\-RTS]

5 = Tripp Lite UPS with Lan2.2 interface (black 73\-0844 cable) 
    [CP=DTR] [OL=CTS] [LB=\-DCD] [SD=DTR+RTS]

6 = Best Patriot with INT51 cable 
    [CP=DTR] [OL=CTS] [LB=\-DCD] [SD=RTS]

7 = CyberPower Power99 
    also Upsonic Power Guardian PG\-500

    Also Belkin Belkin Home Office
    F6H350\-SER, F6H500\-SER, F6H650\-SER

    [CP=RTS] [OL=CTS] [LB=\-DCD] [SD=DTR]

8 = Nitram Elite 500 
    [CP=DTR] [OL=CTS] [LB=\-DCD] [SD=???]

9 = APC Back\(hyUPS/Back\(hyUPS Pro/Smart\(hyUPS with 940\-0023A cable 
    [CP=none] [OL=\-DCD] [LB=CTS] [SD=RTS]

10 = Victron Lite with crack cable 
     [CP=RTS] [OL=CTS] [LB=\-DCD] [SD=DTR]

11 = Powerware 3115 
     [CP=DTR] [OL=\-CTS] [LB=\-DCD] [SD=ST]

12 = APC Back\-UPS Office with 940\-0119A cable 
     [CP=RTS] [OL=\-CTS] [LB=DCD] [SD=DTR]

13 = RPT Repoteck RPT\-800A/RPT\-162A 
     [CP=DTR+RTS] [OL=DCD] [LB=\-CTS] [SD=ST]

14 = Online P\-series 
     [CP=DTR] [OL=DCD] [LB=\-CTS] [SD=RTS]

15 = Powerware 5119, 5125
     [CP=DTR] [OL=CTS] [LB=\-DCD] [SD=ST]

16 = Nitram Elite 2002
     [CP=DTR+RTS] [OL=CTS] [LB=\-DCD] [SD=???]

17 = PowerKinetics 9001
     [CP=DTR] [OL=CTS] [LB=\-DCD] [SD=???]

18 = TrippLite Omni 450LAN with Martin's cabling
     [CP=DTR] [OL=CTS] [LB=DCD] [SD=none]

     http://lists.exploits.org/upsdev/Jul2002/00012.html

19 = Fideltronic Ares Series
     [CP=DTR] [OL=CTS] [LB=\-DCD] [SD=RTS]

20 = Powerware 5119 RM
     [CP=DTR] [OL=\-CTS] [LB=DCD] [SD=ST]

     Check docs/cables/powerware.txt

21 = Generic RUPS 2000 (Megatec M2501 cable)
     [CP=RTS] [OL=CTS] [LB=\-DCD] [SD=RTS+DTR]

     http://lists.exploits.org/upsdev/Oct2004/00004.html

22 = Gamatronic All models with alarm interface
     also CyberPower SL series

     [CP=RTS] [OL=CTS] [LB=\-DCD] [SD=DTR]

.SH SIMILAR MODELS

Many different UPS companies make models with similar interfaces.  The 
RUPS cable seems to be especially popular in the "power strip" variety of 
UPS found in office supply stores.  If your UPS works with an entry in the
table above, but the model or manufacturer information don't match,
don't despair.  You can fix that easily by using the mfr and model 
variables documented above in your \fBups.conf\fR(5).

.SH TESTING COMPATIBILITY

If your UPS isn't listed above, you can try going through the list until 
you find one that works.  There is a lot of cable and interface reuse in 
the UPS world, and you may find a match.

To do this, first make sure nothing important is plugged into the
outlets on the UPS, as you may inadvertently switch it off.  Definitely
make sure that the computer you're using is not plugged into that UPS.
Plug in something small like a lamp so you know when power is being
supplied to the outlets.

Now, you can either attempt to make an educated guess based on the
documentation your manufacturer has provided (if any), or just start
going down the list.  

Step 1

Pick a driver to try from the list (genericups \-h) and go to step 2.

Step 2

Start the driver with the type you want to try \(hy

        genericups \-x upstype=n /dev/port

Let upsd sync up (watch the syslog), and then run upsc to see what it
found.  If the STATUS is right (should be OL for on line), go to step 3,
otherwise go back to step 1.

Alternatively, you can run genericups in debug mode \(hy

        genericups -DDDDD \-x upstype=n /dev/port

In this mode it will be running in the foreground and continuously
display the line and battery status of the UPS.

Step 3

Disconnect the UPS from the wall/mains power.  This is easiest if you
have a switched outlet in between it and the wall, but you can also just
pull the plug to test.  The lamp should stay lit, and the status should
switch to "OB".  If the lamp went out or the status didn't go to "OB"
within about 15 seconds, go to step 1.  Otherwise, continue to step 4.

Step 4

At this point, we know that OL and OB work.  If nothing else beyond
this point works, you at least know what your OL/OB value should be.

Wait for the UPS to start complaining about a low battery.  Depending on
the size of your UPS battery and the lamp's bulb, this could take
awhile.  It should start complaining audibly at some point.  When this
happens, STATUS should show "OB LB" within 15 seconds.  If not, go to
step 1, otherwise continue to step 5.

Step 5

So far: OL works, OB works, and LB works.

With the UPS running on battery, run the genericups driver with the \-k
switch to shut it down.

        genericups \-x upstype=n \-k /dev/port

If the UPS turns off the lamp, you're done.  At this point, you have
verified that the shutdown sequence actually does what you want.  You
can start using the genericups driver with this type number for normal
operations.

You should use your findings to add a section to your ups.conf.
Here is a quick example:

        [myups]
                driver = genericups
                port = /dev/ttyS0
                upstype = 1

Change the port and upstype values to match your system.

.SH NEW SUPPORT

If the above testing sequence fails, you will probably need to create a 
new entry to support your hardware.  All UPS types are determined from the 
table in the genericups.h file in the source tree.

On a standard 9 pin serial port, there are 6 lines that are used as the 
standard "high/low" signal levels.  4 of them are incoming (to the PC, 
from the UPS), and the other 2 are outgoing (to the UPS, from the PC).  
The other 3 are the receive/transmit lines and the ground.

Be aware that many manufacturers remap pins within the cable.  If you have 
any doubts, a quick check with a multimeter should confirm whether the 
cable is straight\(hythrough or not.  Another thing to keep in mind is that 
some cables have electronics in them to do special things.  Some have 
resistors and transistors on board to change behavior depending on what's 
being supplied by the PC.

.SH SPECIFIC MODEL NOTES

These have been contributed by users of this driver.

The Centralion CL series may power down the load if the driver starts up 
with the UPS running on battery as the default line settings contain the 
shutdown sequence.  \(hy Neil Muller

The Tripp\(hyLite Internet Office 700 must be used with the black 73\(hy0844 
cable instead of the gray 73\(hy0743 cable.  This entry should work with any 
of their models with the Lan 2.2 interface \(hy see the sticker by the DB9 
connector on the UPS.  \(hy Stephen Brown

Type 5 should work with the Tripp\(hyLite Lan 2.1 interface and the 73\(hy0724 
cable.  This was tested with the OmniSmart 675 PNP on Red Hat 7.2.  \(hy Q 
Giese

Types 7 and 10 should both work with the PhoenixTec A1000.

.SH BUGS

There is no way to reliably detect a contact\(hyclosure UPS.  This means the 
driver will start up happily even if no UPS is detected.  It also means 
that if the connection between the UPS and computer is interrupted, you 
may not be able to sense this in software.

Most contact\(hyclosure UPSes will not power down the load if the line power 
is present.  This can create a race when using slave \fBupsmon\fR(8) 
systems.  See the \fBupsmon\fR(8) man page for more information.

The solution to both of these problems is to upgrade to a smart protocol 
UPS of some kind that allows detection and proper load cycling on command.

.SH SEE ALSO

.SS The core driver:
\fBnutupsdrv\fR(8)

.SS Internet resources:
The NUT (Network UPS Tools) home page: http://www.networkupstools.org/


NEWS
If you're upgrading from an earlier version, see the UPGRADING file.

---------------------------------------------------------------------------
Release notes for NUT 2.4.1 - what's new since 2.4.0:

 - the microdowell driver has appeared to support various MicroDowell Enterprise
   units (see the "new devices" list below).

 - support for new devices: MicroDowell Enterprise B8, B10, N8, N11, N15, N20,
   N22, N30, N40, N50, N60 and HiBox ST.

 - NUT-Monitor now better handles the ups.status field, and has switched to
   version 1.1.

 - the situation of the build toolchain has been fixed, with regard to the
   "make clean" target and the wrongly removed generated USB files. This brokes
   further configure call. 

---------------------------------------------------------------------------
Release notes for NUT 2.4.0 - what's new since 2.2.2:

 - preliminary support for Power Distribution Units (PDUs): NUT now support
   PDUs, either natively (ie using NUT snmp-ups driver), or through a binding to
   the Powerman daemon. The list of supported PDUs is already quite long,
   including: Eaton ePDUs (Managed and Monitored), some Aphel models, some
   Raritan PDUs, and the whole list of Powerman supported devices:
   http://powerman.sourceforge.net/supported.html
 
 - support for new devices: the various PDUs cited above, Chloride Desk Power
   650, Cyber Power Systems Value 400E/600E/800E (USB models), Delta GES602N,
   Digitus DN-170020, the whole Eaton ranges (mostly composed of MGE Office
   Protection Systems and Powerware units) including BladeUPS, Forza Power
   Technologies SL-1001, HP PowerTrust 2997A, HP R/T 2200 G2, Infosec XP 1000
   and XP 500, Ippon Back Power Pro (serial and USB), Kebo 1200D/D Series,
   Liebert PowerSure Personal XT, MGE Office Protection Systems Protection
   Station, Neus 400va and 600va, Phasak 400VA and 600VA, Plexus 500VA, Powercom
   Black Knight PRO / King PRO and Imperial, PowerKinetics BlackOut Buster,
   Sweex 1000 USB, UNITEK Alpha 500, WinPower CPM-800.

 - NUT now embeds Python client support through the PyNUTClient module and the
   NUT-Monitor application. Both are from David Goncalves, and are still
   available from http://www.lestat.st.
   For more information, refer to scripts/python/README.

 - the dummy-ups driver now support a "repeater" mode. This allows it to act as
   a NUT client, and to forward data. This can be useful for supervision and
   load sharing purposes.

 - tcp-wrappers support has been added to the upsd server, to grant users access
   by source IP for commands that require to be logged into the server. This
   replaces the previous internal implementation (ACL in upsd.conf).

 - the nut.conf file has been introduced to standardize startup configuration
   across the various systems.

 - NUT now ships a bash completion function for 'upsc' command
   (scripts/misc/nut.bash_completion). Simply copy it to /etc/bash_completion.d
 
 - many internal changes to improve maintenability, while lowering the
   maintenance cost (thus allowing developers to focus on what matters: the
   code!). Examples of this are:
   - the USB information automatic extraction to generate the various USB helper
     files,
   - the upsdrv_info_t structure to track more driver information, and remove
     the need for the upsdrv_banner() function
   - common USB code refactoring, as it is done for the serial functions.

 - tons of bugfixes, cleanup and improvements to make NUT stronger than ever!

 For a complete list of changes, please refer to the ChangeLog file.

---------------------------------------------------------------------------
Release notes for NUT 2.2.2 - what's new since 2.2.1:

 - support for new devices: APC BACK-UPS XS LCD, Atlantis Land,
   Mustek Powermust Office 650, Oneac XAU models, Powerware PW5115 and
   PW9120 (USB), Nitram Elite 2005
 
 - Integrated Power Management (NUT HAL integration) has reached a
   major milestone: it is now the most advanced UPS integration into
   Power Management layer known in existing OSs. It has received many
   corrections and improvements, and allows to PowerOff the UPS at the
   end of a power cycle (which is the most important feature, not
   supported on other systems).
   The various files are now installed into the correct location.

 - the usbhid-ups driver has received attention. Most notably, the
   shutdown handling has been reworked, and support for MGE UPS SYSTEMS
   3 phases units has been added.

 - snmp-ups now supports MGE* Environment Sensor (ref 66 846).
   The ambient.temperature reporting has also been fixed for units
   other than APC.

 - the netxml-ups driver has appeared to support MGE* network HTTP/XML
   cards.
 
 - NUT now distributes by default the shared version of libupsclient
   (version 1.0.0), and use this for the provided clients (upsmon, upsc,
   upsrw, upscmd). This is part of an effort to reduce NUT's footprint,
   both on disk and in memory.
 
 - powerpanel has reach a new step toward the replacement of nitram and
   cpsups drivers. The final step is scheduled for NUT 2.4.
 
 - many changes, cleanup and fixes to the NUT core and various drivers.

 For a complete list of changes, please refer to the ChangeLog file.

---------------------------------------------------------------------------
Release notes for NUT 2.2.1 - what's new since 2.2.0:

 - support for new devices: all MGE Office Protection Systems units,
   Advice TopGuard 2000, Belkin F6H375-USB, Dynamix UPS1700D, Effekta RM2000MH,
   Jageson Technology Jasuny USPS, Powercom SMK-1500A and SXL-1500A,
   PowerWalker Line-Interactive VI 400/800 and 600, Powerware 9110,
   UNITEK Alpha 2600, UPSonic CXR1000, some vintage serial APC UPSs.

 - the usbhid-ups driver has been improved, and fixed in many areas, through
   a backport of the development (trunk) version.

 - the udev rules, for Linux hotplug support of the USB UPSs, has been
   updated to support kernel newer than 2.6.22.


drivers/genericups.c
/* genericups.c - support for generic contact-closure UPS models

   Copyright (C) 1999  Russell Kroll <rkroll@exploits.org>

   This program is free software; you can redistribute it and/or modify
   it under the terms of the GNU General Public License as published by
   the Free Software Foundation; either version 2 of the License, or
   (at your option) any later version.

   This program is distributed in the hope that it will be useful,
   but WITHOUT ANY WARRANTY; without even the implied warranty of
   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
   GNU General Public License for more details.

   You should have received a copy of the GNU General Public License
   along with this program; if not, write to the Free Software
   Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA 02111-1307 USA
*/

#include <sys/ioctl.h>

#include "main.h"
#include "serial.h"
#include "genericups.h"

#define DRIVER_NAME	"Generic contact-closure UPS driver"
#define DRIVER_VERSION	"1.36"

/* driver description structure */
upsdrv_info_t upsdrv_info = {
	DRIVER_NAME,
	DRIVER_VERSION,
	"Russell Kroll <rkroll@exploits.org>",
	DRV_STABLE,
	{ NULL }
};

	static	int	upstype = -1;

static void parse_output_signals(const char *value, int *line)
{
	/* parse signals the serial port can output */

	*line = 0;

	if (strstr(value, "DTR") && !strstr(value, "-DTR")) {
		*line |= TIOCM_DTR;
	}

	if (strstr(value, "RTS") && !strstr(value, "-RTS")) {
		*line |= TIOCM_RTS;
	}

	if (strstr(value, "ST")) {
		*line |= TIOCM_ST;
	}

	if (strstr(value, "CTS")) {
		fatalx(EXIT_FAILURE, "Can't override output with CTS (not an output)");
	}

	if (strstr(value, "DCD")) {
		fatalx(EXIT_FAILURE, "Can't override output with DCD (not an output)");
	}

	if (strstr(value, "RNG")) {
		fatalx(EXIT_FAILURE, "Can't override output with RNG (not an output)");
	}

	if (strstr(value, "DSR")) {
		fatalx(EXIT_FAILURE, "Can't override output with DSR (not an output)");
	}
}

static void parse_input_signals(const char *value, int *line, int *val)
{
	/* parse signals the serial port can input */

	*line = 0;
	*val = 0;

	if (strstr(value, "CTS")) {
		*line |= TIOCM_CTS;

		if (!strstr(value, "-CTS")) {
			*val |= TIOCM_CTS;
		}
	}

	if (strstr(value, "DCD")) {
		*line |= TIOCM_CD;

		if (!strstr(value, "-DCD")) {
			*val |= TIOCM_CD;
		}
	}

	if (strstr(value, "RNG")) {
		*line |= TIOCM_RNG;

		if (!strstr(value, "-RNG")) {
			*val |= TIOCM_RNG;
		}
	}

	if (strstr(value, "DSR")) {
		*line |= TIOCM_DSR;

		if (!strstr(value, "-DSR")) {
			*val |= TIOCM_DSR;
		}
	}

	if (strstr(value, "DTR")) {
		fatalx(EXIT_FAILURE, "Can't override input with DTR (not an input)");
	}

	if (strstr(value, "RTS")) {
		fatalx(EXIT_FAILURE, "Can't override input with RTS (not an input)");
	}

	if (strstr(value, "ST")) {
		fatalx(EXIT_FAILURE, "Can't override input with ST (not an input)");
	}
}

void upsdrv_initinfo(void)
{
	char	*v;

	/* setup the basics */

	dstate_setinfo("ups.mfr", "%s", ((v = getval("mfr")) != NULL) ? v : upstab[upstype].mfr);
	dstate_setinfo("ups.model", "%s", ((v = getval("model")) != NULL) ? v : upstab[upstype].model);

	if ((v = getval("serial")) != NULL) {
		dstate_setinfo("ups.serial", "%s", v);
	}

	/*
	 User wants to override the input signal definitions. See also upsdrv_initups().
	 */
	if ((v = getval("OL")) != NULL) {
		parse_input_signals(v, &upstab[upstype].line_ol, &upstab[upstype].val_ol);
		upsdebugx(2, "parse_input_signals: OL overriden with %s\n", v);
	}

	if ((v = getval("LB")) != NULL) {
		parse_input_signals(v, &upstab[upstype].line_bl, &upstab[upstype].val_bl);
		upsdebugx(2, "parse_input_signals: LB overriden with %s\n", v);
	}
}

/* normal idle loop - keep up with the current state of the UPS */
void upsdrv_updateinfo(void)
{
	int	flags, ol, bl, ret;

	ret = ioctl(upsfd, TIOCMGET, &flags);

	if (ret != 0) {
		upslog_with_errno(LOG_INFO, "ioctl failed");
		ser_comm_fail("Status read failed");
		dstate_datastale();
		return;
	}

	ol = ((flags & upstab[upstype].line_ol) == upstab[upstype].val_ol);
	bl = ((flags & upstab[upstype].line_bl) == upstab[upstype].val_bl);

	status_init();

	if (bl) {
		status_set("LB");	/* low battery */
	}

	if (ol) {
		status_set("OL");	/* on line */
	} else {
		status_set("OB");	/* on battery */
	}

	status_commit();

	upsdebugx(5, "ups.status: %s %s\n", ol ? "OL" : "OB", bl ? "BL" : "");

	ser_comm_good();
	dstate_dataok();
}

/* show all possible UPS types */
static void listtypes(void)
{
	int	i;

	printf("Valid UPS types:\n\n");

	for (i = 0; upstab[i].mfr != NULL; i++) {
		printf("%i: %s\n", i, upstab[i].desc);
	}
}

/* set the flags for this UPS type */
static void set_ups_type(void)
{
	int	i;

	if (!getval("upstype")) {
		fatalx(EXIT_FAILURE, "No upstype set - see help text / man page!");
	}

	upstype = atoi(getval("upstype"));

	for (i = 0; upstab[i].mfr != NULL; i++) {

		if (upstype == i) {
			upslogx(LOG_INFO, "UPS type: %s\n", upstab[i].desc);
			return;
		}
	}

	listtypes();

	fatalx(EXIT_FAILURE, "\nFatal error: unknown UPS type number");
}

/* power down the attached load immediately */
void upsdrv_shutdown(void)
{
	int	flags, ret;

	if (upstype == -1) {
		fatalx(EXIT_FAILURE, "No upstype set - see help text / man page!");
	}

	flags = upstab[upstype].line_sd;

	if (flags == -1) {
		fatalx(EXIT_FAILURE, "No shutdown command defined for this model!");
	}

	if (flags == TIOCM_ST) {

#ifndef HAVE_TCSENDBREAK
		fatalx(EXIT_FAILURE, "Need to send a BREAK, but don't have tcsendbreak!");
#endif

		ret = tcsendbreak(upsfd, 4901);

		if (ret != 0) {
			fatal_with_errno(EXIT_FAILURE, "tcsendbreak");
		}

		return;
	}

	ret = ioctl(upsfd, TIOCMSET, &flags);

	if (ret != 0) {
		fatal_with_errno(EXIT_FAILURE, "ioctl TIOCMSET");
	}

	if (getval("sdtime")) {
		int	sdtime;

		sdtime = strtol(getval("sdtime"), (char **) NULL, 10);

		upslogx(LOG_INFO, "Holding shutdown signal for %d seconds...\n",
			sdtime);

		sleep(sdtime);
	}
}

void upsdrv_help(void)
{
	listtypes();
}

void upsdrv_makevartable(void)
{
	addvar(VAR_VALUE, "upstype", "Set UPS type (required)");
	addvar(VAR_VALUE, "mfr", "Override manufacturer name");
	addvar(VAR_VALUE, "model", "Override model name");
	addvar(VAR_VALUE, "serial", "Specify the serial number");
	addvar(VAR_VALUE, "CP", "Override cable power setting");
	addvar(VAR_VALUE, "OL", "Override on line signal");
	addvar(VAR_VALUE, "LB", "Override low battery signal");
	addvar(VAR_VALUE, "SD", "Override shutdown setting");
	addvar(VAR_VALUE, "sdtime", "Hold time for shutdown value (seconds)");
}

void upsdrv_initups(void)
{
	char	*v;

	set_ups_type();

	upsfd = ser_open(device_path);

	/*
	 See if the user wants to override the output signal definitions
	 this must be done here, since we might go to upsdrv_shutdown()
	 immediately. Input signal definition override is handled in
	 upsdrv_initinfo()
	 */
	if ((v = getval("CP")) != NULL) {
		parse_output_signals(v, &upstab[upstype].line_norm);
		upsdebugx(2, "parse_output_signals: CP overriden with %s\n", v);
	}

	if ((v = getval("SD")) != NULL) {
		parse_output_signals(v, &upstab[upstype].line_sd);
		upsdebugx(2, "parse_output_signals: SD overriden with %s\n", v);
	}

	if (ioctl(upsfd, TIOCMSET, &upstab[upstype].line_norm)) {
		fatal_with_errno(EXIT_FAILURE, "ioctl TIOCMSET");
	}
}

void upsdrv_cleanup(void)
{
	ser_close(upsfd, device_path);
}

